### PR TITLE
Add option to r2dbc to close the prepared statement

### PR DESCRIPTION
**Pull request by steveannett**
This pull request would add an option to `JasyncStatement` in jasync-r2dbc-mysql to release the prepared statement after use. By default the prepared statement is left open, but sometimes this is not always the desired option, as reported in https://github.com/jasync-sql/jasync-sql/issues/312

### CODE REVIEW

**Comment by steveannett** [comment]
It should be noted I also submitted a feature request in the `r2dbc-spi` library https://github.com/r2dbc/r2dbc-spi/issues/275

File: r2dbc-mysql/src/main/java/JasyncStatement.kt [hunk 1]
```diff
@@ -26,6 +26,7 @@ internal class JasyncStatement(private val clientSupplier: Supplier<JasyncConnec
 
     private var isPrepared = false
     private var selectLastInsertId: Boolean = false
+    private var releaseAfterwards: Boolean = false
```
Comment (oshai) [inline r2dbc-mysql/src/main/java/JasyncStatement.kt:29]: maybe call it releasePreparedStatementAfterUse (and also the method)?

File: r2dbc-mysql/src/main/java/JasyncStatement.kt [hunk 2]
```diff
@@ -71,6 +72,11 @@ internal class JasyncStatement(private val clientSupplier: Supplier<JasyncConnec
         return this
     }
 
+    fun releaseAfterwards(): Statement {
+        releaseAfterwards = true
```
Comment (oshai) [inline r2dbc-mysql/src/main/java/JasyncStatement.kt:76]: also assert that this is prepared statement?

File: r2dbc-mysql/src/main/java/JasyncStatement.kt [hunk 3]
```diff
@@ -71,6 +72,11 @@ internal class JasyncStatement(private val clientSupplier: Supplier<JasyncConnec
         return this
     }
 
+    fun releaseAfterwards(): Statement {
```
Comment (oshai) [inline r2dbc-mysql/src/main/java/JasyncStatement.kt:75]: how can this be used? it's not part of the interface. by casting to JasyncStatement?

File: r2dbc-mysql/src/main/java/JasyncStatement.kt (hunk 1 above)
Comment (steveannett) [reply r2dbc-mysql/src/main/java/JasyncStatement.kt]: Thanks for the feedback! I've renamed both the variable and method

File: r2dbc-mysql/src/main/java/JasyncStatement.kt (hunk 2 above)
Comment (steveannett) [reply r2dbc-mysql/src/main/java/JasyncStatement.kt]: I've added a check that it is a prepared statement, let me know if that is OK, or if you'd prefer it in a different style

File: r2dbc-mysql/src/main/java/JasyncStatement.kt (hunk 3 above)
Comment (steveannett) [reply r2dbc-mysql/src/main/java/JasyncStatement.kt]: Yes, that would be the idea, if the programmer was really keen on making sure that the prepared statement is closed they could cast it and request for it to be closed, for example in Java 14:

```
	private void releaseAfterUse(Statement st) {
		if (st instanceof JasyncStatement jst) jst.releasePreparedStatementAfterUse();
	}
```

**Comment by oshai** [comment]
basically LGTM. One more idea about configuring it: how about adding an option to `MysqlConnectionFactoryProvider`?

I am not a spring expert but maybe that is the intended usage for spring? I guess we can have both to allow connection level control as well. WDYT?

File: r2dbc-mysql/src/main/java/JasyncStatement.kt (hunk 2 above)
Comment (oshai) [reply r2dbc-mysql/src/main/java/JasyncStatement.kt]: lgtm

**Comment by oshai** [comment]
@steveannett is it ready to be merged?

**Comment by steveannett** [comment]
@oshai yes, ready to be merged, thanks!

Regarding adding an option to `MySqlConnectionFactoryProvider`, its a good idea, I will look into this and see if I can get it added - although in my particular use case I would like to have the connection level control so I can leave it open in certain cases.

**Pull request merged**